[TIR] Implement TIR macros

kparzysz-quic opened the pull request:
This patch introduces new symbol: `T.macro`. `T.macro` is a decorator that, when applied to a function, turns the body of that function into a piece of TIR that can be inserted into a PrimFunc by calling the macro the same way as a function.

For example:

```python
@T.macro
def copy_backwards(dst, src, size):
    with T.block("backwards"):
        for i in T.serial(size):
            ai = T.axis.remap("S", [i])
            T.reads(src[0:size])
            T.writes(dst[0:size])
            dst[ai] = src[size - ai - 1]

@T.prim_func
def foo_int32(A: T.Buffer((128,), "int32"), B: T.Buffer((128,), "int32")):
    copy_backwards(A, B, 128)

@T.prim_func
def foo_int8(A: T.Buffer((128,), "int8"), B: T.Buffer((128,), "int8")):
    copy_backwards(A, B, 128)
```

The above will generate two PrimFuncs that do the same backwards copy, but applied to buffers with different data types.

```python
@T.prim_func
def foo_int32(A: T.Buffer((128,), "int32"), B: T.Buffer((128,), "int32")):
    # with T.block("root"):
    for i in range(128):
        with T.block("copy"):
            ai = T.axis.spatial(128, i)
            T.reads(B[0:128])
            T.writes(A[0:128])
            A[ai] = B[128 - ai - 1]

@T.prim_func
def foo_int8(A: T.Buffer((128,), "int8"), B: T.Buffer((128,), "int8")):
    # with T.block("root"):
    for i in range(128):
        with T.block("copy"):
            ai = T.axis.spatial(128, i)
            T.reads(B[0:128])
            T.writes(A[0:128])
            A[ai] = B[128 - ai - 1]
```

Semantics:
- Function that is decorated with @T.macro can have any parameters that follow Python syntax, i.e. positional, keyword, etc. Type annotations are not required, but are allowed.
- For `macro_name(arg1, arg2, arg3, ...)`, the values are substituted into the body of the macro, and the body with the substituted values is then inserted at the point where the call is located.
- Macros are hygienic by default, but can be made non-hygienic via a keyword argument `hygienic` to `T.macro`, e.g. `T.macro(hygienic=False)`.

Lunderberg commented on python/tvm/script/parser/tir/entry.py:
```diff
@@ -50,6 +50,48 @@ def prim_func(func: Callable) -> Union[PrimFunc, Callable]:
+    # This function is a decorator, so the return value will replace
+    # the function definition (to which the decorator it is applied)
+    # in that function's name space.
+    return obj
```
We should collect any closure variables that are used within the function, so that they can be later be added to the macro's variables.  This would be similar to how the `@T.prim_func` and `@I.ir_module` decorators capture the closure variables for use in metaprogramming.

```python
obj.__closure_vars = utils.inspect_function_capture(func)
```

kparzysz-quic replied:
Done.

Lunderberg reviewed (commented):
I like it, just some questions on simplifying the implementation, which variables should be in-scope within the body of the macro, and whether macros should be provided for use in Relax as well as TIR.

I also liked your point from the discuss thread about the root `tir::Block` being handled differently if trying to generate a `Stmt` externally, and agree that it is better to re-parse the body of the macro in order to avoid those special cases.

kparzysz-quic commented on python/tvm/script/parser/tir/parser.py:
```diff
@@ -528,3 +543,76 @@ def visit_tvm_declare_function(self: Parser, node: doc.FunctionDef) -> GlobalVar
+        self.report_error(node, f"Undefined macro '{macro_name}'")
+
+    if isinstance(macro.doc, doc.Module):
+        macro_def = find_macro_def(macro_name, macro.doc.body)
```
In my tests it was always a Module...  I just added the other cases because I thought that they may happen.  Would you expect FunctionDef instead?

Lunderberg replied:
Huh, looks like I was entirely off about the return type of `ast.parse`.  I had expected it to return the most specific AST type that can describe the string, but even if I just run `ast.parse("1 + 2")` it results in an `ast.Module`.  Checking [here](https://docs.python.org/3/library/ast.html#ast.Module), it looks like that's guaranteed behavior, so we couldn't get a `doc.FunctionDef` at this location anyways.

kparzysz-quic reviewed (commented):
Thanks for the review!

kparzysz-quic commented on python/tvm/script/parser/tir/entry.py:
```diff
@@ -50,6 +50,48 @@ def prim_func(func: Callable) -> Union[PrimFunc, Callable]:
+
+def macro(func: Callable) -> doc.AST:
+    obj = TIRMacro(*parse_macro(func))
+    setattr(obj, "__name__", func.__name__)
```
I followed the code from `prim_func` above.  You're right, the `setattr` is unnecessary.

kparzysz-quic replied:
Done.

kparzysz-quic commented on python/tvm/script/parser/tir/parser.py:
```diff
@@ -427,6 +428,20 @@ def visit_expr_stmt(self: Parser, node: doc.Expr) -> None:
         The doc AST Expr node.
     """
+
+    def is_insert_macro(node: doc.Call) -> bool:
```
Done.

kparzysz-quic commented on python/tvm/script/parser/tir/parser.py:
```diff
@@ -427,6 +428,20 @@ def visit_expr_stmt(self: Parser, node: doc.Expr) -> None:
+        attr = node.func
+        if not isinstance(attr.value, doc.Name):
+            return False
+        if attr.value.id != "T" or attr.attr != "insert":
```
Done.

kparzysz-quic replied:
`T.insert` has been removed.

kparzysz-quic commented on python/tvm/script/parser/tir/entry.py:
```diff
@@ -50,6 +50,48 @@ def prim_func(func: Callable) -> Union[PrimFunc, Callable]:
+def macro(func: Callable) -> doc.AST:
+    obj = TIRMacro(*parse_macro(func))
+    setattr(obj, "__name__", func.__name__)
+    # We don't need to explicitly store the return value anywhere.
```
Done.

kparzysz-quic commented on python/tvm/script/parser/tir/parser.py:
```diff
@@ -528,3 +543,76 @@ def visit_tvm_declare_function(self: Parser, node: doc.FunctionDef) -> GlobalVar
+    # with it.
+
+    # Construct the function with the macro's parameters, and returning locals().
+    macro_ast = doc.from_doc(macro_def)
```
Done.

kparzysz-quic commented on python/tvm/script/parser/tir/entry.py:
```diff
@@ -50,6 +50,48 @@ def prim_func(func: Callable) -> Union[PrimFunc, Callable]:
+#   where the `T.insert` is located.
+
+
+class TIRMacro:
```
Would it be `I.macro` then?

Lunderberg replied:
It would, yes.  Within TVMScript, the default set of global definitions is defined [here](https://github.com/apache/tvm/blob/main/python/tvm/script/parser/core/entry.py#L44).  This provides both `tvm.script.tir` as `T` and `tvm.script.ir` as `I`.

kparzysz-quic replied:
Can we do this in a separate PR?

kparzysz-quic commented on python/tvm/script/parser/tir/parser.py:
```diff
@@ -528,3 +543,76 @@ def visit_tvm_declare_function(self: Parser, node: doc.FunctionDef) -> GlobalVar
+    exec(cmacro, self.var_table.get(), local_vars)  # pylint: disable=exec-used
+    local_vars = local_vars[tmp_name]
+
+    with self.var_table.with_frame():
```
Done.

kparzysz-quic commented:
Do you have any more comments? @Lunderberg

Lunderberg reviewed (approved):
Looks good, and thank you for making the changes!

I think the only open question was which namespace the macro support should go under.  By placing it under the `tvm.script.tir` namespace, we'll need to keep a backwards-compatibility import if/when we move it to `tvm.script.ir`, unless done in the current PR.  That said, that would be a pretty useful import anyways, regardless of the backwards compatibility, so I'm not worried there.

junrushao commented:
Didn't have the chance to review timely but this looks really awesome to me!

kparzysz-quic commented:
@Lunderberg, @junrushao 
I'm starting to work on making this more general and extending it to Relax.  Relax code has a lot of assignments, the targets of which would disappear from the frame in the current implementation of `expand_macro`.  My first thought was to do something like
```
@R.macro(outputs=["var2, "var3"]):
def something():
   var1 = ...
   var2 = ...
   var3 = ...

@R.function
def something_else(...)
   something()
   ... = var3
```
Admittedly, `var3 = something()` looks a lot prettier, but it's sort of breaks the idea behind a macro: the macro would have to "return" a value, which implies a lot more complications in parsing or expanding it.

I'm wondering if you have any better ideas.

Lunderberg commented:
In Relax, I think this use case could be handled by a function that returns a `relax.Expr`.  Unlike in TIR, Relax has an additional normalization step that flattens out any `relax::SeqExpr`.  So, if you had thje function `something()` return a `relax::SeqExpr`, then call it with `var3 = something()`, I think that would result in the behavior you want.

For multiple return values, I can't remember if Relax's normalizer would remove trivial Tuple bindings, where the function return value is a tuple which gets immediately unpacked, but that would be a reasonable extension to add to it.

kparzysz-quic commented:
For the purposes of the rest of the comment, I renamed `TIRMacro` to `ScriptMacro`, and created a subclass `RelaxMacro`.

There are some complications with macros returning values, specifically when the call to the macro is a part of a larger expression.  The problem is that there are no specific parser visitors for subexpressions---the whole expression is evaluated by the evaluator all at once, so I can't do the same "AST injection" as I did for TIR (at least not without adding detailed visitors and capturing AST.Call nodes).

For example,
```
@R.macro
def m():
    return 1

@R.function
def foo():
    x = m() + 1
```
The entire `m() + 1` will be evaluated, causing Python to actually call the `RelaxMacro` object.  I went down that route a bit, added `__call__` operator to `RelaxMacro`, parsed it as if it was `R.function`, and returned the body.  This works, but has severe limitations: the macro's parameters need to be annotated and can only be objects that a `R.function` can take, so no `str` for example.

I'm inclined to follow the TIR route instead, but I wanted to consult with you in case this is a bad idea.

kparzysz-quic commented:
I think I got it to work:
```python
import tvm
from tvm import relax
from tvm.script import relax as R

@R.macro
def alloc_and_shape(dtype: str):
    alloc = R.builtin.alloc_tensor(R.shape([4, 4]), runtime_device_index=0, dtype=dtype)
    shape = R.shape_of(alloc)
    return shape

@R.function
def foo(x: R.Tensor((4, 4), "float32")):
    shape = alloc_and_shape(dtype="float32")
    return shape


print(alloc_and_shape)
print()
print(foo)
```

Produces
```python
@R.macro
def alloc_and_shape(dtype: str):
    alloc = R.builtin.alloc_tensor(R.shape([4, 4]), runtime_device_index=0, dtype=dtype)
    shape = R.shape_of(alloc)
    return shape


# from tvm.script import relax as R

@R.function
def foo(x: R.Tensor((4, 4), dtype="float32")) -> R.Shape([4, 4]):
    alloc: R.Tensor((4, 4), dtype="float32") = R.builtin.alloc_tensor(R.shape([4, 4]), R.dtype("float32"), R.prim_value(0))
    shape: R.Shape([4, 4]) = R.shape_of(alloc)
    shape_1: R.Shape([4, 4]) = shape
    return shape_1
```
